clients/src/main/java/org/apache/kafka/common/KafkaFuture.java

```diff
@@ -31,7 +31,7 @@
      * A function which takes objects of type A and returns objects of type B.
      */
     public static abstract class Function<A, B> {
-        public abstract B apply(A a);
+        public abstract B apply(A a) throws Throwable;
```

Comment:
That's a good point.  It might be good to avoid throwing checked exceptions from Function#apply, to make it easier to transition to Java 8 when it becomes available.  Note that we already have a slight incompatibility here since in Java 8, throwing an exception E from Function#apply results in Future#get throwing a CompletionException that wraps E.  However, since CompletionException is not present in Java 7, KafkaFuture#get will give you back an ExecutionException that wraps E instead.